### PR TITLE
Calendar Report

### PR DESCRIPTION
**Pull request by lelemm**
A new report type where data shows in a calendar format`:

https://github.com/user-attachments/assets/229e42e9-3b74-435f-9b00-92c14a273e8a

### CODE REVIEW

File: packages/desktop-client/src/components/reports/reportRanges.ts
Comment (lelemm) [inline packages/desktop-client/src/components/reports/reportRanges.ts:1]: This is a fix because when offset is one, its returning 2 months instead of one.

File: packages/desktop-client/src/components/transactions/TransactionsTable.jsx
Comment (lelemm) [reply packages/desktop-client/src/components/transactions/TransactionsTable.jsx:1523]: out of the scope of the PR

File: packages/desktop-client/src/components/reports/graphs/CalendarGraph.tsx
Comment (lelemm) [reply packages/desktop-client/src/components/reports/graphs/CalendarGraph.tsx:114]: done

File: packages/desktop-client/src/components/reports/spreadsheets/calendar-spreadsheet.ts
Comment (lelemm) [reply packages/desktop-client/src/components/reports/spreadsheets/calendar-spreadsheet.ts:39]: this looks incomplete, the variable "filters" is used on line 64, and its not inside the try

File: packages/desktop-client/src/style/themes/development.ts
Comment (lelemm) [reply packages/desktop-client/src/style/themes/development.ts:217]: they are used for different calendars

File: packages/desktop-client/src/components/reports/reports/CalendarCard.tsx
Comment (lelemm) [reply packages/desktop-client/src/components/reports/reports/CalendarCard.tsx:367]: debouncedResizeCallback is used with useResizeObserver:

import { useRef, useCallback } from 'react';

export function useResizeObserver(
  func: (contentRect: DOMRectReadOnly) => void,
): (el: unknown) => void {
  const observer = useRef(null);
  if (!observer.current) {
    observer.current = new ResizeObserver(entries => {
      func(entries[0].contentRect);
    });
  }

  const elementRef = useCallback(el => {
    observer.current.disconnect();
    if (el) {
      observer.current.observe(el, { box: 'border-box' });
    }
  }, []);

  return elementRef;
}

can you expand your logic so I can understand what you mean

**Comment by joel-jeremy** [comment]
Some observations:
- Does this need its own experimental feature flag?
  - Currently if the customizable dashboards flag is off, there is no way to create a calendar report
- There is a bug where if you keep on swiping up the collapsible transaction list, it also keeps on moving up

**Comment by lelemm** [comment]
> Could the dashboard tooltips say the month instead of just "Day"? So "Nov 13" instead of "Day 13" image

I did that based on the monthly spending.
image
I will change to Month + day

**Comment by lelemm** [comment]
done @youngcw. 
Limited to 4, then it adds a scrollbar, also changed the tooltip title

**Comment by youngcw** [comment]
The scroll bar seems a bit thick, but other than that looks great!

**Comment by lelemm** [comment]
> The scroll bar seems a bit thick, but other than that looks great!

that's the default horizontal scrollbar style:
image
image

**Comment by youngcw** [comment]
The popovers for the calendar dashboard card are different than the other reports.  But to be honest using that style looks nicer.  Maybe the others should be updated instead of this changed.

When selecting different days from the full report, the calendar at the top will move up and down since not all days have both income and spending.  It would be nice if that stayed consistent and had the same sizing no matter if the days had both in and out, or just one.

It would also be nice if you could select a range of days when looking at the full report.  Maybe the ability to ctrl-click and shift-click on days to make selections and ranges.  That may be for a different PR if that is too much for now.

If may be nice from a UX view to unselect the selected day in the full report if you click outside the calendar.

**Comment by lelemm** [comment]
> The popovers for the calendar dashboard card are different than the other reports. But to be honest using that style looks nicer. Maybe the others should be updated instead of this changed.

Yeah, the other reports use recharts component, the tooltip is part of the component, the tooltip i'm using here is the standard actual tooltip component. That's why its different.

> 
> When selecting different days from the full report, the calendar at the top will move up and down since not all days have both income and spending. It would be nice if that stayed consistent and had the same sizing no matter if the days had both in and out, or just one.

I think I don't get this one. Can you record/screenshot what you mean?

> 
> It would also be nice if you could select a range of days when looking at the full report. Maybe the ability to ctrl-click and shift-click on days to make selections and ranges. That may be for a different PR if that is too much for now.
> 

That's a nice idea, but I guess It would take rework to make this feasible, may be another PR would be better. 

> It may be nice from a UX view to unselect the selected day in the full report if you click outside the calendar. Right now you have to remove the filter that gets generated.

yeah, nice one. I will do it

**Comment by youngcw** [comment]
This is what I mean.  The calendar shifts around as you click different days

https://github.com/user-attachments/assets/e04a6971-af29-4c19-93c5-f4cba7a45518

**Comment by lelemm** [comment]
> This is what I mean. The calendar shifts around as you click different days
> 
>  cal.mp4

got it

**Comment by lelemm** [comment]
> > It may be nice from a UX view to unselect the selected day in the full report if you click outside the calendar. Right now you have to remove the filter that gets generated.
> 
> yeah, nice one. I will do it

turns out, not easy, lol. I will check if I can find a nice way to do this

**Comment by Jonathan-Fang** [comment]
This is a great report to add! Looking forward to the end product!

File: packages/desktop-client/src/components/reports/graphs/CalendarGraph.tsx
Comment (joel-jeremy) [inline packages/desktop-client/src/components/reports/graphs/CalendarGraph.tsx:305]: Would `AutoTextSize` component work here?

File: packages/desktop-client/src/components/reports/graphs/CalendarGraph.tsx
Comment (joel-jeremy) [inline packages/desktop-client/src/components/reports/graphs/CalendarGraph.tsx:277]: We should just use `className` + emotion's `css` here to merge classnames and styles in one declaration

File: packages/desktop-client/src/components/reports/reports/Calendar.tsx
Comment (joel-jeremy) [inline packages/desktop-client/src/components/reports/reports/Calendar.tsx:133]: Can we use `useEffect` here?

File: packages/desktop-client/src/components/reports/reports/Calendar.tsx
Comment (joel-jeremy) [inline packages/desktop-client/src/components/reports/reports/Calendar.tsx:365]: We should declare these outside the component

File: packages/desktop-client/src/components/reports/reports/Calendar.tsx
Comment (joel-jeremy) [inline packages/desktop-client/src/components/reports/reports/Calendar.tsx:349]: Should event handler methods in this component be wrapped in useCallback?

File: packages/desktop-client/src/components/reports/reports/Calendar.tsx
Comment (joel-jeremy) [inline packages/desktop-client/src/components/reports/reports/Calendar.tsx:371]: We should put the 115 in a variable with meaningful name. Is it an offset?

File: packages/desktop-client/src/components/reports/reports/Calendar.tsx
Comment (joel-jeremy) [inline packages/desktop-client/src/components/reports/reports/Calendar.tsx:394]: Same comment above, should these be wrapped in useCallback?

File: packages/desktop-client/src/components/reports/reports/Calendar.tsx
Comment (joel-jeremy) [inline packages/desktop-client/src/components/reports/reports/Calendar.tsx:559]: Do we actually need Promise.resolve here or can we just return an empty array?

File: packages/desktop-client/src/components/reports/reports/Calendar.tsx
Comment (joel-jeremy) [inline packages/desktop-client/src/components/reports/reports/Calendar.tsx:691]:
```suggestion
type CalendarWithHeaderProps = {
```

File: packages/desktop-client/src/components/reports/reports/Calendar.tsx
Comment (joel-jeremy) [inline packages/desktop-client/src/components/reports/reports/Calendar.tsx:822]: Does View not work here?

File: packages/desktop-client/src/components/reports/reports/CalendarCard.tsx
Comment (joel-jeremy) [inline packages/desktop-client/src/components/reports/reports/CalendarCard.tsx:46]:
```suggestion
type CalendarCardProps = {
```

File: packages/desktop-client/src/components/reports/reports/CalendarCard.tsx
Comment (joel-jeremy) [inline packages/desktop-client/src/components/reports/reports/CalendarCard.tsx:362]: Would useMemo work here?

File: packages/desktop-client/src/components/reports/graphs/CalendarGraph.tsx
Comment (lelemm) [reply packages/desktop-client/src/components/reports/graphs/CalendarGraph.tsx:305]: image
it would need a rework on the rest of the component to make it fit. its very big
but, after adjusting it each number will get its own size making it inconsistent sizing

File: packages/desktop-client/src/components/reports/reports/Calendar.tsx
Comment (lelemm) [reply packages/desktop-client/src/components/reports/reports/Calendar.tsx]: changed

File: packages/desktop-client/src/components/reports/reports/CalendarCard.tsx
Comment (lelemm) [reply packages/desktop-client/src/components/reports/reports/CalendarCard.tsx]: changed to useMemo, it works properly

File: packages/desktop-client/src/components/reports/reports/Calendar.tsx
Comment (lelemm) [reply packages/desktop-client/src/components/reports/reports/Calendar.tsx:827]: It works with display 'block', changed to it

File: packages/desktop-client/src/components/reports/reports/Calendar.tsx
Comment (lelemm) [reply packages/desktop-client/src/components/reports/reports/Calendar.tsx]: done, its the size of the summary, so the expandable transactions don't hide it

File: packages/desktop-client/src/components/reports/reports/Calendar.tsx
Comment (lelemm) [reply packages/desktop-client/src/components/reports/reports/Calendar.tsx]: done

File: packages/desktop-client/src/components/reports/reports/Calendar.tsx
Comment (lelemm) [reply packages/desktop-client/src/components/reports/reports/Calendar.tsx]: done

File: packages/desktop-client/src/components/reports/graphs/CalendarGraph.tsx
Comment (lelemm) [reply packages/desktop-client/src/components/reports/graphs/CalendarGraph.tsx:281]: this is a left out from testing classes, removed the classnames

**Comment by lelemm** [comment]
> > > It may be nice from a UX view to unselect the selected day in the full report if you click outside the calendar. Right now you have to remove the filter that gets generated.
> > 
> > 
> > yeah, nice one. I will do it
> 
> turns out, not easy, lol. I will check if I can find a nice way to do this

done @youngcw 


> This is looking very good! Thank you for your time and effort on this!
ty  @joel-jeremy  :)
> 
> I do have some comments on mobile view:
> 
> 1. When opening the filter in mobile, the whole page seems to re-render and the filter tooltip automatically closes
I noticed that too, I have to investigate what is going on. Any suspicion?

> 2. When the transaction list/drawer is open and you swipe down on the top of it, the drawer moves down to where your finger stops. I think swiping down should fully close the drawer.

changed how it works, take a look

> 3. Can we make the `Show transactions`/`Hide transactions` button larger so it's easier to click? Please also make it sentence case since we use sentence case throughout the app.
change the case and size of the button

> 4. Would it be easy to make the calendars line up vertically on mobile instead of horizontally so that we can utilize more screen real estate? (flexDirection = column for mobile?)
atm the calendar direction depends on how it shows on the desktop version, if height > width -> flex direction = column, otherwise, flex direction = row:
https://github.com/user-attachments/assets/e5034700-939a-4871-8575-c137d2b161f3





> Should we still display the Income and Expense summary as 0 when there are none for the selected days? image

removed the conditions to hide when zero

**Comment by youngcw** [comment]
My only nitpik now is that the whole page shifts down when you add in a filter for the first time, which is a bit jarring.  Thats not that big of a deal for now though.

**Comment by lelemm** [comment]
> My only nitpik now is that the whole page shifts down when you add in a filter for the first time, which is a bit jarring. Thats not that big of a deal for now though.

added a minHeight to the header, so as long as there is not too much filters, it will keep the screen static

File: packages/desktop-client/src/components/reports/reports/Calendar.tsx
Comment (joel-jeremy) [inline packages/desktop-client/src/components/reports/reports/Calendar.tsx:658]: LGTM! One last nitpick, Can we remove the press highlight on this button? The text disappears when the button is clicked :)

Comment (lelemm) [reply packages/desktop-client/src/components/reports/reports/Calendar.tsx:658]: do you mean like this?
image

Comment (joel-jeremy) [reply packages/desktop-client/src/components/reports/reports/Calendar.tsx:659]: Yup, I think that would work :)

Comment (lelemm) [reply packages/desktop-client/src/components/reports/reports/Calendar.tsx:659]: I committed it already.

what's up with this action?
Compare Sizes / compare (pull_request_target) Failing after 21s
mind explaining me when it fails?

Comment (matt-fidd) [reply packages/desktop-client/src/components/reports/reports/Calendar.tsx:659]: It looked to be a flake, I've rerun and it passed now

**Review by joel-jeremy — APPROVED** [review]
LGTM! Would prefer another maintainer to approve before merging.

I think the next phase for the calendar report is to include preview transactions i.e. https://github.com/actualbudget/actual/issues/1928. Not sure if we want to do it here or in a separate PR. I'm fine either ways. :)

**Comment by youngcw** [comment]
This needs the same update as the summary card to allow imnporting a dashboard.json file

**Comment by lelemm** [comment]
> This needs the same update as the summary card to allow importing a dashboard.json file

I'm not sure what we need to do here. can you clarify please?

**Comment by youngcw** [comment]
> > This needs the same update as the summary card to allow importing a dashboard.json file
> 
> I'm not sure what we need to do here. can you clarify please?

You need to copy what Mike did on the summary report

**Comment by lelemm** [comment]
> > > This needs the same update as the summary card to allow importing a dashboard.json file
> > 
> > 
> > I'm not sure what we need to do here. can you clarify please?
> 
> You need to copy what Mike did on the summary report

ohh, got it!

done ✔️

**Comment by lelemm** [comment]
@youngcw done. I guess I missed one filter =p

File: packages/desktop-client/src/components/reports/reports/Calendar.tsx
Comment (matt-fidd) [inline packages/desktop-client/src/components/reports/reports/Calendar.tsx:618]:
```suggestion
                        <Trans>No transactions</Trans>
```

File: packages/desktop-client/src/components/reports/reports/Calendar.tsx
Comment (matt-fidd) [inline packages/desktop-client/src/components/reports/reports/Calendar.tsx:675]:
```suggestion
                        <Trans>Show transactions</Trans>
```

File: packages/desktop-client/src/components/reports/reports/Calendar.tsx
Comment (matt-fidd) [inline packages/desktop-client/src/components/reports/reports/Calendar.tsx:681]:
```suggestion
                        <Trans>Hide transactions</Trans>
```

File: packages/desktop-client/src/components/reports/reports/CalendarCard.tsx
Comment (matt-fidd) [inline packages/desktop-client/src/components/reports/reports/CalendarCard.tsx:219]:
```suggestion
                            <Trans>Income:</Trans>
```

File: packages/desktop-client/src/components/reports/reports/CalendarCard.tsx
Comment (matt-fidd) [inline packages/desktop-client/src/components/reports/reports/CalendarCard.tsx:11]:
```suggestion
import { Trans, useTranslation } from 'react-i18next';
```

File: packages/desktop-client/src/components/reports/reports/CalendarCard.tsx
Comment (matt-fidd) [inline packages/desktop-client/src/components/reports/reports/CalendarCard.tsx:240]:
```suggestion
                            <Trans>Expenses:</Trans>
```

**Review by matt-fidd — CHANGES_REQUESTED** [review]
Just a couple of missed translations, everything else looks good

**Comment by matt-fidd** [comment]
I just noticed the table is missing the parent transaction

Calendar:
image


Ledger:
image

**Comment by lelemm** [comment]
> I just noticed the table is missing the parent transaction
> 
> Calendar: image
> 
> Ledger: image

any ideas how/why?

**Comment by matt-fidd** [comment]
> > I just noticed the table is missing the parent transaction
> > Calendar: image
> > Ledger: image
> 
> any ideas how/why?

Sorry, I'm not sure - it looks like the parent just isn't in the data

File: packages/desktop-client/src/components/reports/reports/CalendarCard.tsx
Comment (lelemm) [reply packages/desktop-client/src/components/reports/reports/CalendarCard.tsx:219]: @jfdoming how would we do the privacy filter on this example?
`<Trans>Income: {{ incomeValue }}</Trans>`

Comment (jfdoming) [reply packages/desktop-client/src/components/reports/reports/CalendarCard.tsx:219]: I think `<Trans>Income: <Privacy filter>{{ incomeValue }}</PrivacyFilter></Trans>` should work

Comment (jfdoming) [reply packages/desktop-client/src/components/reports/reports/CalendarCard.tsx:219]: And generate a string like "Income: <1>{incomeValue}</1>"

**Comment by lelemm** [comment]
> > > I just noticed the table is missing the parent transaction
> > > Calendar: image
> > > Ledger: image
> > 
> > 
> > any ideas how/why?
> 
> Sorry, I'm not sure - it looks like the parent just isn't in the data

fixed

File: packages/desktop-client/src/components/reports/reports/CalendarCard.tsx
Comment (lelemm) [reply packages/desktop-client/src/components/reports/reports/CalendarCard.tsx:219]: @jfdoming  I kept @matt-fidd suggestion since Income and the value are not even on the same level:
image

It may open for confusing context translations, but It keeps the layout as it is

**Review by matt-fidd — APPROVED** [review]
LGTM, will leave to get another one/two reviews before merging

Great addition!

**Review by youngcw — APPROVED** [review]

**Review by jfdoming — APPROVED** [review]

File: packages/desktop-client/src/components/reports/reports/CalendarCard.tsx
Comment (jfdoming) [reply packages/desktop-client/src/components/reports/reports/CalendarCard.tsx]: I see. Maybe in the future we may need to update, but for now I guess this is fine

**Pull request merged**
